multi_attribute_toydown/toydown.py

```diff
@@ -0,0 +1,340 @@
+        """ Adds Laplacian noise of parameter 1/`epsilon` to Node `node`. If `epsilon` is np.inf, 
+            adds no noise.  `epsilon` must be positive, otherwise throws ValueError
+        """
+        if epsilon <= 0: raise ValueError("epsilon <= 0")
+
+        if self.gurobi:
+             node.data.noise = {}
+             node.data.noised = {}
+             
+             for k, v in node.data.attributes.items():
+                shape = v.shape
+                noise = np.random.laplace(scale=1/epsilon, size=shape)
+                node.data.noise[k] = noise
+                node.data.noised[k] = v + noise
+
+        else:
+            shape = node.data.attributes.shape
+            noise = np.random.laplace(scale=1/epsilon, size=shape)
+            node.data.noised = node.data.attributes + noise
+            node.data.noise = noise
+        
+        node.data.noise_type = "laplacian"
+
+    def add_levels_to_node(self, node, level):
+        """ Recursive function that adds the `level` as an attribute
+            to the `node`.
+        """
+        self.get_node(node.identifier).data.level = level
+        for child in self.children(node.identifier):
+            self.add_levels_to_node(child, level+1)
+
+    def noise_and_adjust(self, objective_fun="L1", node_cons=None, bounds="non-negative", 
+                         parental_equality=True, maxiter=200, verbose=False, save=False):
+        """ Noises each node in the Tree and adjusts them to add back to their parent.
+            This function simply serves as a wrapper function to the recursive 
+            __noise_and_adjust_children(), and is started at the root of the tree.
+
+            objective_fun     : Function to minimize over.  If "L1" uses the L1 norm, 
+                                if "L2" uses the L2 norm, otherwise a function that takes a
+                                noised point, and returns a function that takes a point x and returns 
+                                a scalar distance between the two points.
+            node_cons         : Function that takes the number of children, n, and returns a list of constraints 
+                                for the readjustment of a Node is subject to. 
+                                Has form [{"type": "eq/ineq", "fun": lambda x: }]
+                                See scipy.optimize constraint specification for more details.
+            bounds            : Function that takes the number of children, n, and returns a list of 
+                                (min, max) pairs for each attribute in each of the children, defining 
+                                the bounds on that attribute. Use None for one of min or max when there is 
+                                no bound in that direction.
+                                A value of "non-negative", flags that all counts should be > 0
+            parental_equality : Boolean flag - adds constraint that for all attributes, the sum of the 
```

Comment:
Is there ever a case when we don't want the children to add up to the parents?

Reply:
In general no there won't be.  I put the flag in there mostly for debugging, but also so that if we wanted to look at the error of individual constraints we could toggle all of them.